atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/CassandraGetCandidateCellsForSweepingImpl.java

```diff
@@ -0,0 +1,155 @@
+                    for (CassandraRawCellValue col : cols) {
+                        processColumn(col, candidates);
+                    }
+                }
+                if (candidates.isEmpty()) {
+                    end = true;
+                    if (!currentTimestamps.isEmpty()) {
+                        return ImmutableList.of(createCandidate());
+                    } else {
+                        return endOfData();
+                    }
+                } else {
+                    return candidates;
+                }
+            }
+        }
+
+        private void processColumn(CassandraRawCellValue col, @Output List<CandidateCellForSweeping> candidates) {
+            Pair<byte[], Long> colNameAndTs = CassandraKeyValueServices.decomposeName(col.getColumn());
+            Cell cell = Cell.create(col.getRowKey(), colNameAndTs.getLhSide());
+            if (!cell.equals(currentCell)) {
+                if (!currentTimestamps.isEmpty()) {
+                    candidates.add(createCandidate());
+                }
+                currentCell = cell;
+            }
+            long ts = colNameAndTs.getRhSide();
+            if (ts < request.sweepTimestamp()) {
+                if (currentTimestamps.isEmpty()) {
+                    // Timestamps are in the decreasing order, so we pick the first timestamp below sweepTimestamp
+                    // to check the value for emptiness
+                    currentLatestValEmpty = request.shouldCheckIfLatestValueIsEmpty()
+                                        && col.getColumn().getValue().length == 0;
+                }
+                currentTimestamps.add(ts);
+                numCellTsPairsExamined += 1;
+            }
+        }
+
+        private CandidateCellForSweeping createCandidate() {
+            boolean isCandidate = isCandidate();
+            long[] sortedTimestamps = isCandidate ? sortTimestamps() : EMPTY_LONG_ARRAY;
+            currentTimestamps.clear();
+            return ImmutableCandidateCellForSweeping.builder()
+                    .cell(currentCell)
+                    .sortedTimestamps(sortedTimestamps)
+                    .isLatestValueEmpty(currentLatestValEmpty)
+                    .numCellsTsPairsExamined(numCellTsPairsExamined)
+                    .build();
+        }
+
+        private long[] sortTimestamps() {
+            currentTimestamps.reverse();
+            return currentTimestamps.toArray();
+        }
+
+        private boolean isCandidate() {
+            return currentTimestamps.size() > 1
+                    || currentLatestValEmpty
+                    || (currentTimestamps.size() == 1 && timestampIsPotentiallySweepable(currentTimestamps.get(0)));
+        }
+
+        private boolean timestampIsPotentiallySweepable(long ts) {
+            return ts == Value.INVALID_VALUE_TIMESTAMP || ts >= request.minUncommittedStartTimestamp();
+        }
+    }
+
+    private static final long[] EMPTY_LONG_ARRAY = new long[0];
```

Comment:
We tend to put constants up at the top of the class, rather than hidden away at the bottom.